react/features/subtitles/middleware.js

```diff
@@ -0,0 +1,113 @@
+
+    switch (action.type) {
+    case ENDPOINT_MESSAGE_RECEIVED:
+        return _endpointMessageReceived(store, next, action);
+    }
+
+    return next(action);
+});
+
+/**
+ * Notifies the feature transcription that the action
+ * {@code ENDPOINT_MESSAGE_RECEIVED} is being dispatched within a specific redux
+ * store.
+ *
+ * @param {Store} store - The redux store in which the specified {@code action}
+ * is being dispatched.
+ * @param {Dispatch} next - The redux {@code dispatch} function to
+ * dispatch the specified {@code action} to the specified {@code store}.
+ * @param {Action} action - The redux action {@code ENDPOINT_MESSAGE_RECEIVED}
+ * which is being dispatched in the specified {@code store}.
+ * @private
+ * @returns {Object} The value returned by {@code next(action)}.
+ */
+function _endpointMessageReceived({ dispatch, getState }, next, action) {
+    const json = action.json;
+
+    try {
+
+        // Let's first check if the given object has the correct
+        // type in the json, which identifies it as a json message sent
+        // from Jigasi with speech-to-to-text results
+        if (json.type === 'transcription-result') {
+            // Extract the useful data from the json.
+            const isInterim = json.is_interim;
+            const participantName = json.participant.name;
+            const stability = json.stability;
+            const text = json.transcript[0].text;
+            const transcriptMessageID = json.message_id;
+
+            // If this is the first result with the unique message ID,
+            // we add it to the state along with the name of the participant
+            // who said given text
+            if (!getState()['features/subtitles']
+                .transcriptMessages.has(transcriptMessageID)) {
+                dispatch(addTranscriptMessage(transcriptMessageID,
+                    participantName));
+            }
+            const { transcriptMessages }
+                = getState()['features/subtitles'];
+            const newTranscriptMessage
+                = transcriptMessages.get(transcriptMessageID);
+
+            // If this is final result, update the state as a final result
+            // and start a count down to remove the subtitle from the state
+            if (!isInterim) {
+
+                newTranscriptMessage.final = text;
+                dispatch(updateTranscriptMessage(transcriptMessageID,
+                    newTranscriptMessage));
+
+                setTimeout(() => {
+                    dispatch(removeTranscriptMessage(transcriptMessageID));
+                }, REMOVE_AFTER_MS);
+            } else if (stability > 0.85) {
+
+                // If the message has a high stability, we can update the
+                // stable field of the state and remove the previously
+                // unstable results
+
+                newTranscriptMessage.stable = text;
+                newTranscriptMessage.unstable = undefined;
```

Comment:
Is there any danger of modifying the object in place instead of creating a copy of newTranscriptMessage?

Reply:
Doesn't seem to cause any trouble but I will make a copy as it is redux recommends not to update state variable directly and reducer states should be immutable.